### ISSUE TITLE
[mypyc] property setter and deleter don't work?

### ISSUE THREAD

**Issue by nstarman** [python/mypy#13231]
**Bug Report**

(A clear and concise description of what the bug is.)

**To Reproduce**

1. Defined (in a file `foo.py`) a class `Bar` with a property that has a setter and deleter.
2. run ``mypyc foo.py``3. 

e.g.

```python
class Bar:
    _attr: int | None
    def __init__(self) -> None:
        self._attr = None
    @property
    def attr(self) -> int:
        return self._attr
    @attr.setter
    def attr(self, value: int) -> None:
        self._attr = value
    @attr.deleter
    def attr(self) -> None:
        self._attr = None
```

**Expected Behavior**

<!—
  How did you expect your project to behave?
  It’s fine if you’re not sure your understanding is correct.
  Write down what you thought would happen. If you just expected no errors, you can delete this section.
—>

`mypyc` would compile.

**Actual Behavior**

<!—
  Did something go wrong?
  Is something broken, or not behaving as you expected?
—>

I get the following error

```
File "mypyc/irbuild/classdef.py", line 87, in transform_class_def
File "mypyc/irbuild/classdef.py", line 200, in add_method
File "mypyc/irbuild/function.py", line 369, in handle_ext_method
File "mypyc/irbuild/function.py", line 312, in gen_func_item
File "mypyc/irbuild/function.py", line 351, in gen_func_ir
...: KeyError: <mypy.nodes.FuncDef object at 0x7f93d4876b80>
```

**Your Environment**

<!-- Include as many relevant details about the environment you experienced the bug in —>

- Mypy version used: 0.950
- Mypy command-line flags: N/A
- Mypy configuration options from `mypy.ini` (and other config files): N/A
- Python version used: 3.9
- Operating system and version: macOS Monterey, version 12.4

<!—
You can freely edit this text, please remove all the lines
you believe are unnecessary.
—>

**Comment by nstarman** [comment]
@AlexWaygood, I narrowed the source of the problem. It isn't setters and deleters by themselves, but in conjunction with (at least) weakref (and maybe other types).

**Comment by nstarman** [comment]
Just tested on 

```python
from __future__ import annotations
from typing import Generic, Protocol, TypeVar


T = TypeVar("T")
O = TypeVar("O", covariant=True)


class Thing(Protocol[O]):
    def __call__(self) -> O: ...


class Bar(Generic[T]):
    _attr: Thing[T] | None
    def __init__(self) -> None:
        self._attr = None
    @property
    def attr(self) -> T | None:
        return self._attr() if self._attr is not None else None
    @attr.deleter  # compiles if deleter (and what it decorates) is commented out
    def attr(self) -> None:
        self._attr = None
```

This raises the same exception, so I suspect the problem is actually with setting parameterized attributes?

**Comment by AlexWaygood** [comment]
I'm afraid I'm only a triager here, and mypyc is one of the areas of mypy I know least about, so I'm probably not the person you want to be tagging here, I'm afraid :)

**Comment by nstarman** [comment]
Whom should I tag?

**Comment by AlexWaygood** [comment]
97littleleaf11, ichard26 or JukkaL might be interested. (ichard26 isn't a core dev, but he's made several mypyc contributions and uses mypyc for compiling `black`.)

But we're all volunteers with limited free time :)

**Comment by nstarman** [comment]
Thanks @AlexWaygood.

@JukkaL, I think I've found an issue with how mypyc processes parametrized attributes, at least in the context of decorated methods. I've been trying to understand the error trace to submit a PR resolving this issue, but I don't yet grok Mypy(s)'s internals. Hopefully this and #13304 are easily resolved issues.
